### ISSUE TITLE
Proposal: Experimental features interface / API

### ISSUE THREAD

**Issue by loujaybee** [gitpod-io/gitpod#7026]
This issue is a proposal to implement for a product-wide approach for experimental features in Gitpod. 

## Context / Motivations
We already have the concept of "VS Code Insiders" which is an experimental pre-release designed for power-users to be able to test out features in VS Code before they're rolled out to the full user base (see below) [[1](https://gitpod.io/preferences)]. 

<img src="https://user-images.githubusercontent.com/5528307/144410056-d797a65f-6c41-45b5-9df3-6376cbf8cb62.png" width="400" />

But the IDE interface is now starting to get cluttered, and the distinction between pre-release versions and stable will be less clear (it already was not entirely clear what the insider's version is). We are now considering applying the same approach (insiders/early access IDE's) for EAP [[1](https://www.jetbrains.com/resources/eap/)] versions of JetBrains IDE's, where we have a pre-release version and a stable. Additionally, we've had challenges in the past of users not knowing when a feature is a beta / pre-release vs a stable one. This causes potential to upset or confusion for users and causes additional support burden due to increased GitHub tickets, questions from users, etc. 

The docs labels help somewhat, but do not guarantee that a user understands fully the implications of an experimental feature: 
![image](https://user-images.githubusercontent.com/5528307/144410954-a10e4d05-b7cc-4046-9887-3b9b23662290.png)

We have beta labels in-app, but again, do not ensure that a user is entirely aware of implications of an experimental feature:
![image](https://user-images.githubusercontent.com/5528307/144410971-ae7e8678-1d84-4e2b-af8d-445d99617a28.png)

## Solution

* A clear description of the experimental feature (to the user) e.g. what the feature does, what are the impacts
* A way to toggle on/off the feature

From a technical design perspective, the solution needs to be able to be easily plugged into other aspects of the application, so a logical API and/or interface is essential. The solution should also play nicely with any existing feature flag behaviours, and potentially could be used for A/B testing in the future. 

Related [internal Slack discussion](https://gitpod.slack.com/archives/C01KGM9BH54/p1638442840357200?thread_ts=1638436313.350800&cid=C01KGM9BH54).

**Comment by loujaybee** [comment]
Relates to #7081

**Comment by loujaybee** [comment]
CC @mikenikles — it would be useful if this could also apply to changes in the docs, as I'm thinking about how we ensure docs updates are synced also with feature updates, I believe you were thinking about moving docs into the main repo somehow so that we can review both at the same time?

**Comment by loujaybee** [comment]
Note: It has been discussed elsewhere to give community heroes some forms of early access to beta features, maybe this feature can help facilitate that, also.

**Comment by pawlean** [comment]
> Note: It has been discussed elsewhere to give community heroes some forms of early access to beta features, maybe this feature can help facilitate that, also.

This sounds awesome!

**Comment by mikenikles** [comment]
Yes, I have an experimental PR at https://github.com/gitpod-io/website/pull/1367.

The technical limitation we had in April no longer exists. There are a few tasks listed in the PR that need to happen first though.
